python/tests/test_diffracted_planewave.py

```diff
@@ -137,12 +139,102 @@ def _pw_amp(x):
 
     print("PASSED.")
 
+
   def test_diffracted_planewave(self):
     self.run_binary_grating_diffraction(2.6,0.4,0.6,0)
     self.run_binary_grating_diffraction(2.6,0.4,0.6,13.4)
 
     # self.run_binary_grating_diffraction(10.0,0.5,0.5,0)
     # self.run_binary_grating_diffraction(10.0,0.5,0.5,10.7)
 
+
+  def run_mode_source(self,gp,gh,gdc,m,diffpw):
+    """Computes the transmitted flux of a
+       binary grating given an incident planewave
+       specified by the diffraction order `m` in the
+       y direction. The incident planewave is defined
+       using a mode source with either a band number
+       or `DiffractedPlanewave` object specified by
+       the boolean flag `diffpw`.
+    """
+    sx = self.dpml+self.dsub+gh+self.dpad+self.dpml
+    sy = gp
+    cell_size = mp.Vector3(sx,sy,0)
+
+    ky = m/gp
+    theta = math.asin(ky/(self.fcen*self.ng))
+
+    # k (in source medium) with correct length (plane of incidence: XY)
+    k = mp.Vector3(self.fcen*self.ng).rotate(mp.Vector3(z=1), theta)
+
+    eig_parity = mp.ODD_Z
+    if theta == 0:
+      k = mp.Vector3()
+      eig_parity += mp.EVEN_Y
+      symmetries = [mp.Mirror(direction=mp.Y)]
+    else:
+      symmetries = []
+
+    src_pt = mp.Vector3(-0.5*sx+self.dpml,0,0)
+    if diffpw:
+      # the *zeroth* diffraction order defines a planewave with a
+      # wavevector equal to the `k_point` of the `Simulation` object
+      sources = [mp.EigenModeSource(mp.GaussianSource(self.fcen,fwidth=0.1*self.fcen),
+                                    center=src_pt,
+                                    size=mp.Vector3(0,sy,0),
+                                    eig_band=mp.DiffractedPlanewave((0,0,0),
+                                                                    mp.Vector3(0,1,0),
+                                                                    1,
+                                                                    0))]
+    else:
+      sources = [mp.EigenModeSource(mp.GaussianSource(self.fcen,fwidth=0.1*self.fcen),
+                                    center=src_pt,
+                                    size=mp.Vector3(0,sy,0),
+                                    direction=mp.NO_DIRECTION,
+                                    eig_kpoint=k,
```

Comment:
```suggestion
```
Should be able to delete these lines as I commented in #2072.  (It's possible you were having problems before that are now fixed by #2114.)

Reply:
If it's not working, make sure the MPB/Meep  verbosity is > 1 and look at the [`NEW KPOINT`](https://github.com/NanoComp/meep/blob/5aac79a41c84cb14a64a844379036a12c14fc51d/src/mpb.cpp#L516) and [`MPB solved for frequency`](https://github.com/NanoComp/meep/blob/5aac79a41c84cb14a64a844379036a12c14fc51d/src/mpb.cpp#L571-L573) verbose outputs — you should see that it is setting the parallel component of `k` by the boundary conditions, and varying the perpendicular component of `k`.

Reply:
Looks like there is a potential bug because printing out `NEW KPOINT` shows that its value is clearly incorrect:
```
NEW KPOINT: 3, 2, 0
```
The *y* component (i.e., the periodic direction of the monitor monitor) should be 1.333... (i.e., 4/3) to match the `k_point` of the `Simulation` object. The reason it is not this value is due to these lines:

https://github.com/NanoComp/meep/blob/405b7e607f87d7dd4a81baf3d5473aa4c28e7a4a/src/mpb.cpp#L426-L433

In this example, `R = (1,1.5,1)` and `kcart = (0,1.33333,0)` which results in `k = (0, 2.0, 0)`.

The fix seems to be to just replace `k[i] = dot_product(R[i], kcart)` with `k[i] = kcart[i]`?

Reply:
No, that code is correct — MPB expects the k point in the reciprocal-lattice basis, so it has to convert from the Cartesian basis of Meep.  (It does the same thing with the `DiffractedPlanewave` object.)

Instead, something weird is going on.  You said the output is:
```
MPB solved for frequency_1(3,1.33333,0) = 2 after 29 iters
```
which means that it *is* taking the nonzero ky into account, but somehow it is getting the wrong kx — somehow, it is acting as though it is passing `ky==0` to MPB, even though we are clearly printing that it is passing a nonzero `k[1]`.

Reply:
I think I have finally gotten to the bottom of this issue. TLDR: the current behavior is a *feature* and not a bug.

Currently, to launch oblique planewaves using an `EigenModeSource` with `eig_band` set to an integer (rather than an `DiffractedPlanewave` object), the user must provide its wavevector `k` (a `meep.Vector3` object) via `direction=mp.NO_DIRECTION` *and* `eig_kpoint=k`. This is because even though internally Meep sets the component of `k` parallel to the line source (in the directions with periodic boundaries) to the components of the `k_point` of the `Simulation` object, the correct perpendicular component $\left|k_\perp\right| = (n\omega)^2 - \left|k_\parallel\right|$ is [*only computed automatically for the case when `eig_band` is a `DiffractedPlanewave` object*](https://github.com/NanoComp/meep/blob/a93a1585b0bd4b3525423c91ed88064d4d54c2da/src/mpb.cpp#L612-L659). 

It would be nice to extend this feature to when `eig_band` is an integer but that can be a separate PR since it is not directly related to this PR.

Reply:
No.  A `DiffractedPlanewave` object only enables the *analytical* calculation of k.  Otherwise, it uses a Newton solver to find the k where ω(k) = specified frequency.  This of course should be the same thing (but slower, and with more ambiguity about polarization and more difficulty in choosing a specific diffraction order).

Reply:
The wavevector of the oblique planewave in Cartesian coordinates should be: `k = (2.68742, 1.33333, 0.00000)`.  However, the Newton solver shows:
```
MPB solved for frequency_1(3,1.33333,0) = 2 after 29 iters
update_maxwell_data_k:, k=(3.00000, 2.00000, 0.00000)
update_maxwell_data_k:, G[0]=(1.00000, 0.00000, 0.00000)
update_maxwell_data_k:, G[1]=(0.00000, 0.66667, 0.00000)
update_maxwell_data_k:, G[2]=(0.00000, 0.00000, 1.00000)
```

Note that the $y$ component (parallel direction) of the MPB mode (`1.33333`) is correct but the $x$ component (perpendicular direction) (`3`) is not `2.68742`.

Reply:
$|k|=n\omega=(1.5)(2)=3$ but clearly the wavevector from the  Newton solver is wrong since $|3,1.3333,0)| > 3$.

Reply:
Turns out that there was a bug in the *output* of MPB's mode wavevector which is fixed by #2253. (The mode itself was correct.)

With the bug fix, the output from `run_mode_source` of `test_diffracted_planewave.py` in this PR is:
```
MPB solved for frequency_1(2.68742,1.33333,0) = 2 after 23 iters
```
This means `k = (2.68742,1.33333,0)` which is equivalent to the `k_point` of the `Simulation` class object as expected.